Add IPv[4|6] multicast join/leave via socket API

jukkar opened the pull request:
Zephyr has its own multicast join/leave API but for interoperability, it should be possible to use the socket API.
So add support for IP_ADD_MEMBERSHIP and IP_DROP_MEMBERSHIP for IPv4 multicast groups. and IPV6_ADD_MEMBERSHIP and IPV6_DROP_MEMBERSHIP for IPv6 multicast groups.
Add also tests for these four.

jukkar committed:
net: socket: Add IPv4 multicast join/leave via socket

Zephyr has its own multicast join/leave API but for
interoperability, it is possible to use the multicast
socket API and IP_ADD_MEMBERSHIP and IP_DROP_MEMBERSHIP
socket options.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: igmp: Add socket based join/leave tests

Make sure the IP_ADD_MEMBERSHIP and IP_DROP_MEMBERSHIP
work as expected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: socket: Add IPv6 multicast join/leave via socket

Zephyr has its own multicast join/leave API but for
interoperability, it is possible to use the multicast
socket API and IPV6_ADD_MEMBERSHIP and IPV6_DROP_MEMBERSHIP
socket options.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
tests: net: mld: Add socket based join/leave tests

Make sure the IPV6_ADD_MEMBERSHIP and IPV4_DROP_MEMBERSHIP
work as expected.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

pdgendt reviewed (dismissed):
This is really useful!

One question: if the socket is bound to the unspecified IP address, it will probably pick the first interface, is there a way to join on all interfaces in that case?
Or doesn't it work like that? :)

rlubos commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -2590,6 +2592,46 @@ int z_vrfy_zsock_getsockopt(int sock, int level, int optname,
 #include <syscalls/zsock_getsockopt_mrsh.c>
 #endif /* CONFIG_USERSPACE */
 
+static int ipv4_multicast_group(struct net_context *ctx, const void *optval,
```
The `ctx` argument is unused, maybe remove it from this static function?

jukkar replied:
Actually I think we can leave the ctx there as it can be used to select an interface if user has not supplied one. I will add support for this in next version.

rlubos replied:
Ok!

rlubos commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -2632,6 +2633,40 @@ static int ipv4_multicast_group(struct net_context *ctx, const void *optval,
 	return 0;
 }
 
+static int ipv6_multicast_group(struct net_context *ctx, const void *optval,
```
Ditto, `ctx` is unused.

jukkar commented:
> One question: if the socket is bound to the unspecified IP address, it will probably pick the first interface, is there a way to join on all interfaces in that case?
> Or doesn't it work like that? :)

The linux manual page for IPv4 says this
```
       imr_address is the address of the local interface with which the
       system should join the multicast group; if it is equal to
       INADDR_ANY, an appropriate interface is chosen by the system.
       imr_ifindex is the interface index of the interface that should
       join/leave the imr_multiaddr group, or 0 to indicate any
       interface.
```
So if I am reading that correctly, it is not possible to join all the interfaces in one go.

jukkar commented:
> So if I am reading that correctly, it is not possible to join all the interfaces in one go.

In a way, the index 0 could be considered to mean "all interfaces" as it kind of makes sense. Anyway, we can think about that in another PR when/if needed.

henrikbrixandersen reviewed (dismissed):
Thanks. Please add an entry in the release notes for this.

jukkar commented:
> Please add an entry in the release notes for this.

Usually these minor enhancements are added when the release is finalized but sure we can add stuff in the PR too.
@henrikbrixandersen could you merge #65886 while I am adding the release note, as it will conflict with this one and I can then rebase this on top of that.

jukkar committed:
doc: release: 3.6: Add info for IPv4/6 multicast socket options

Add information about the IPv4 multicast IP_ADD_MEMBERSHIP and
IP_DROP_MEMBERSHIP socket options.
Add information about the IPv6 multicast IPV6_ADD_MEMBERSHIP and
IPV6_DROP_MEMBERSHIP socket options.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

henrikbrixandersen reviewed (commented):
+1 for relnotes. Thanks.
